pkg/protocols/grpcprot/request.go

```diff
@@ -0,0 +1,306 @@
+	// 1.in the case of replication, the dst object should be consistent with the source object,
+	// so all modifications should be written directly into source object's ctx
+	// 2.the atomicity of header.
+	Request struct {
+		stream   grpc.ServerStream
+		ctx      context.Context
+		cancel   context.CancelFunc
+		header   *Header
+		headerMx *sync.Mutex
+		realIP   string
+		sts      *serverTransportStream
+		peer     *peer.Peer
+	}
+
+	serverTransportStream struct {
+		grpc.ServerTransportStream
+		path string
+	}
+	// Addr is a net.Addr backed by either a TCP "ip:port" string, or
+	// the empty string if unknown.
+	Addr struct {
+		addr string
+	}
+)
+
+const (
+	Authority = ":authority"
+)
+
+var (
+	_ protocols.Request = (*Request)(nil)
+)
+
+// NewRequestWithServerStream creates a new request from a grpc.ServerStream
+// that have data and conn, so it could reply to client with data
+func NewRequestWithServerStream(stream grpc.ServerStream) *Request {
+	r := &Request{
+		stream:   stream,
+		headerMx: &sync.Mutex{},
+	}
+	r.ctx, r.cancel = context.WithCancel(r.stream.Context())
+
+	r.sts = &serverTransportStream{ServerTransportStream: grpc.ServerTransportStreamFromContext(r.ctx)}
+	// grpc's method equals request.path in http standard lib
+	if method, ok := grpc.Method(r.ctx); ok {
+		r.sts.path = method
+	}
+	r.ctx = grpc.NewContextWithServerTransportStream(r.ctx, r.sts)
+	md, ok := metadata.FromIncomingContext(r.ctx)
+	if !ok {
+		logger.Infof("couldn't get headers' md from grpc context")
+		md = metadata.New(nil)
+	}
+
+	r.header = NewHeader(md)
+	r.ctx = metadata.NewIncomingContext(r.ctx, r.header.md)
+	r.peer = &peer.Peer{Addr: &Addr{}}
+
+	if peerInfo, ok := peer.FromContext(r.ctx); ok {
+		r.peer.Addr.(*Addr).setAddr(peerInfo.Addr.String())
+		r.peer.AuthInfo = peerInfo.AuthInfo
+	}
+	r.ctx = peer.NewContext(r.ctx, r.peer)
+
+	return r
+}
+
+// NewRequestWithContext creates a new request from context.
+// that only have request data, not support reply to client
+func NewRequestWithContext(ctx context.Context) *Request {
+	r := &Request{
+		ctx:      ctx,
+		headerMx: &sync.Mutex{},
+	}
+	r.ctx, r.cancel = context.WithCancel(r.ctx)
+
+	r.sts = &serverTransportStream{ServerTransportStream: grpc.ServerTransportStreamFromContext(r.ctx)}
+	// grpc's method equals request.path in http standard lib
+	if method, ok := grpc.Method(r.ctx); ok {
+		r.sts.path = method
+	}
+	r.ctx = grpc.NewContextWithServerTransportStream(r.ctx, r.sts)
+	md, ok := metadata.FromIncomingContext(r.ctx)
+	if !ok {
+		logger.Infof("couldn't get headers' md from grpc context")
+		md = metadata.New(nil)
+	}
+
+	r.header = NewHeader(md)
+	r.ctx = metadata.NewIncomingContext(r.ctx, r.header.md)
+	r.peer = &peer.Peer{Addr: &Addr{}}
+
+	if peerInfo, ok := peer.FromContext(r.ctx); ok {
+		r.peer.Addr.(*Addr).setAddr(peerInfo.Addr.String())
+		r.peer.AuthInfo = peerInfo.AuthInfo
+	}
+	r.ctx = peer.NewContext(r.ctx, r.peer)
+
+	return r
```

Comment:
these lines of code are duplicated with the above function, please do a refactor.